File: src/test/java/no/nav/sbl/dialogarena/mininnboks/provider/rest/henvendelse/HenvendelseControllerTest.java
package no.nav.sbl.dialogarena.mininnboks.provider.rest.henvendelse;

import no.nav.brukerdialog.security.context.SubjectRule;
import no.nav.brukerdialog.security.domain.IdentType;
import no.nav.common.auth.SsoToken;
import no.nav.common.auth.Subject;
import no.nav.sbl.dialogarena.mininnboks.TestUtils;
import no.nav.sbl.dialogarena.mininnboks.consumer.HenvendelseService;
import no.nav.sbl.dialogarena.mininnboks.consumer.TekstService;
import no.nav.sbl.dialogarena.mininnboks.consumer.domain.*;
import no.nav.sbl.dialogarena.mininnboks.consumer.utils.HenvendelsesUtils;
import no.nav.sbl.dialogarena.mininnboks.provider.rest.henvendelse.HenvendelseController.NyHenvendelseResultat;
import no.nav.tjeneste.domene.brukerdialog.henvendelse.v1.sendinnhenvendelse.meldinger.WSSendInnHenvendelseResponse;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.mockito.stubbing.Answer;

import javax.ws.rs.core.Response;
import javax.xml.ws.soap.SOAPFaultException;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static java.util.Arrays.asList;
import static java.util.Collections.*;
import static java.util.stream.Collectors.toList;
import static no.nav.sbl.dialogarena.mininnboks.TestUtils.now;
import static no.nav.sbl.dialogarena.mininnboks.TestUtils.nowPlus;
import static org.apache.commons.lang3.StringUtils.join;
import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.*;


@RunWith(MockitoJUnitRunner.class)
public class HenvendelseControllerTest {
    @Mock
    HenvendelseService service;

    @Mock
    TekstService tekstService = mock(TekstService.class);

    @InjectMocks
    HenvendelseController controller = new HenvendelseController();

    @Rule
    public SubjectRule subjectRule = new SubjectRule(new Subject("fnr", IdentType.EksternBruker, SsoToken.oidcToken("token", emptyMap())));

    @Before
    public void setup() {
        final List<Henvendelse> henvendelser = asList(
                new Henvendelse("1").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(now()),
                new Henvendelse("2").withTraadId("2").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(now()),
                new Henvendelse("3").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(now()),
                new Henvendelse("4").withTraadId("3").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(now()),
                new Henvendelse("5").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(now()),
                new Henvendelse("6").withTraadId("2").withType(Henvendelsetype.SPORSMAL_MODIA_UTGAAENDE).withOpprettetTid(nowPlus(100)),
                new Henvendelse("7").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(now())
        );

        HenvendelsesUtils.setTekstService(tekstService);

        when(service.hentAlleHenvendelser(anyString())).thenReturn(henvendelser);

        when(service.hentTraad(anyString())).thenAnswer((Answer<List<Henvendelse>>) invocation -> {
            String traadId = (String) invocation.getArguments()[0];

            return henvendelser.stream()
                    .filter(henvendelse -> traadId.equals(henvendelse.traadId))
                    .collect(toList());
        });

        when(service.sendSvar(any(Henvendelse.class), anyString())).thenReturn(
                new WSSendInnHenvendelseResponse().withBehandlingsId(UUID.randomUUID().toString())
        );
    }

    @After
    public void after() {
        HenvendelsesUtils.setTekstService(null);
    }

    @Test
    public void henterUtAlleHenvendelserOgGjorOmTilTraader() throws Exception {
        List<Traad> traader = controller.hentTraader();
        assertThat(traader.size(), is(3));
    }

    @Test
    public void filtrererBortUavsluttedeDelsvar() {
        when(service.hentAlleHenvendelser(anyString())).thenReturn(mockBehandlingskjedeMedDelsvar());

        List<Traad> traader = controller.hentTraader();
        Optional<Henvendelse> delsvar = traader.get(0).meldinger.stream()
                    .filter(henvendelse -> henvendelse.type == Henvendelsetype.DELVIS_SVAR_SKRIFTLIG)
                    .findAny();

        assertThat(delsvar.isPresent(), is(false));
    }

    private List<Henvendelse> mockBehandlingskjedeMedDelsvar() {
        return Arrays.asList(
                new Henvendelse("123").withType(Henvendelsetype.SPORSMAL_SKRIFTLIG).withTraadId("1").withOpprettetTid(now()),
                new Henvendelse("234").withType(Henvendelsetype.DELVIS_SVAR_SKRIFTLIG).withTraadId("1").withOpprettetTid(now())
        );
    }

    @Test
    public void serviceKanFeileUtenAtEndepunktFeiler() throws Exception {
        when(service.hentAlleHenvendelser(anyString())).thenReturn(emptyList());
        List<Traad> traader = controller.hentTraader();
        assertThat(traader.size(), is(0));
    }

    @Test
    public void henterUtEnkeltTraadBasertPaId() throws Exception {
        Traad traad1 = (Traad) controller.hentEnkeltTraad("1").getEntity();
        Traad traad2 = (Traad) controller.hentEnkeltTraad("2").getEntity();
        Traad traad3 = (Traad) controller.hentEnkeltTraad("3").getEntity();

        assertThat(traad1.meldinger.size(), is(4));
        assertThat(traad2.meldinger.size(), is(2));
        assertThat(traad3.meldinger.size(), is(1));
    }

    @Test
    public void henterUtTraadSomIkkeFinnes() throws Exception {
        Response response = controller.hentEnkeltTraad("avabv");

        assertThat(response.getStatus(), is(404));
    }

    @Test
    public void girStatuskodeIkkeFunnetHvisHenvendelseServiceGirSoapFault() {
        when(service.hentTraad(anyString())).thenThrow(SOAPFaultException.class);

        Response response = controller.hentEnkeltTraad("1");

        assertThat(response.getStatus(), is(Response.Status.NOT_FOUND.getStatusCode()));
    }

    @Test
    public void markeringSomLest() throws Exception {
        controller.markerSomLest("1");

        verify(service, times(1)).merkSomLest("1");
    }

    @Test
    public void markeringAlleSomLest() throws Exception {
        controller.markerAlleSomLest("1");

        verify(service, times(1)).merkAlleSomLest("1");
    }

    @Test
    public void kanIkkeSendeSvarNarSisteHenvendelseIkkeErSporsmal() throws Exception {
        Svar svar = new Svar();
        svar.traadId = "1";
        svar.fritekst = "Tekst";

        Response response = controller.sendSvar(svar);

        assertThat(response.getStatus(), is(Response.Status.NOT_ACCEPTABLE.getStatusCode()));
    }

    @Test
    public void kanSendeSvarNarSisteHenvendelseErSporsmal() throws Exception {
        Svar svar = new Svar();
        svar.traadId = "2";
        svar.fritekst = "Tekst";


        NyHenvendelseResultat nyHenvendelseResultat = ((NyHenvendelseResultat) controller.sendSvar(svar).getEntity());

        assertThat(nyHenvendelseResultat.behandlingsId, is(not(nullValue())));
    }

    @Test
    public void kopiererNyesteErTilknyttetAnsattFlaggTilSvaret() {
        Henvendelse henvendelse1 = new Henvendelse("1");
        henvendelse1.erTilknyttetAnsatt = true;
        henvendelse1.opprettet = now();
        henvendelse1.type = Henvendelsetype.SPORSMAL_MODIA_UTGAAENDE;
        Henvendelse henvendelse2 = new Henvendelse("2");
        henvendelse2.erTilknyttetAnsatt = false;
        henvendelse2.opprettet = now();
        List<Henvendelse> henvendelser = asList(henvendelse1, henvendelse2);
        when(service.hentTraad(anyString())).thenReturn(henvendelser);

        Svar svar = new Svar();
        svar.fritekst = "fritekst";
        svar.traadId = "0";
        controller.sendSvar(svar);

        ArgumentCaptor<Henvendelse> henvendelseArgumentCaptor = ArgumentCaptor.forClass(Henvendelse.class);
        verify(service).sendSvar(henvendelseArgumentCaptor.capture(), anyString());

        assertThat(henvendelseArgumentCaptor.getValue().erTilknyttetAnsatt, is(true));
    }

    @Test
    public void kopiererBrukersEnhetTilSvaret() {
        String brukersEnhet = "1234";

        Henvendelse henvendelse1 = new Henvendelse("1");
        henvendelse1.opprettet = nowPlus(-1);
        henvendelse1.type = Henvendelsetype.SPORSMAL_SKRIFTLIG;
        henvendelse1.brukersEnhet = brukersEnhet;
        Henvendelse henvendelse2 = new Henvendelse("2");
        henvendelse2.opprettet = now();
        henvendelse2.type = Henvendelsetype.SPORSMAL_MODIA_UTGAAENDE;
        when(service.hentTraad(anyString())).thenReturn(asList(henvendelse1, henvendelse2));

        Svar svar = new Svar();
        svar.fritekst = "fritekst";
        svar.traadId = "0";
        controller.sendSvar(svar);

        ArgumentCaptor<Henvendelse> henvendelseArgumentCaptor = ArgumentCaptor.forClass(Henvendelse.class);
        verify(service).sendSvar(henvendelseArgumentCaptor.capture(), anyString());

        assertThat(henvendelseArgumentCaptor.getValue().brukersEnhet, is(brukersEnhet));
    }

    @Test
    public void senderIkkeFunnetNaarTraadOptionalIkkeErPresent() {
        when(service.hentTraad(anyString())).thenReturn(null);

        Svar svar = new Svar();
        svar.fritekst = "fritekst";
        svar.traadId = "0";
        Response response = controller.sendSvar(svar);

        assertThat(response.getStatus(), is(Response.Status.NOT_FOUND.getStatusCode()));

    }

    @Test(expected = AssertionError.class)
    public void smellerHvisTomFritekstISporsmal() {
        Sporsmal sporsmal = new Sporsmal();
        sporsmal.fritekst = "";
        sporsmal.temagruppe = Temagruppe.ARBD.name();
        controller.sendSporsmal(sporsmal);
    }

    @Test(expected = AssertionError.class)
    public void smellerHvisForLangFritekstISporsmal() {
        Sporsmal sporsmal = new Sporsmal();
        sporsmal.fritekst = join(nCopies(1001, 'a'), "");
        sporsmal.temagruppe = Temagruppe.ARBD.name();
        controller.sendSporsmal(sporsmal);
    }

    @Test(expected = AssertionError.class)
    public void smellerHvisAndreSosialtjenesterTemagruppeISporsmal() {
        Sporsmal sporsmal = new Sporsmal();
        sporsmal.fritekst = "DUMMY";
        sporsmal.temagruppe = Temagruppe.ANSOS.name();
        controller.sendSporsmal(sporsmal);
    }
}

File: src/test/java/no/nav/sbl/dialogarena/mininnboks/provider/rest/henvendelse/HenvendelseControllerTest.kt
package no.nav.sbl.dialogarena.mininnboks.provider.rest.henvendelse

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.whenever
import io.ktor.application.install
import io.ktor.features.ContentNegotiation
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.jackson.JacksonConverter
import io.ktor.routing.routing
import io.ktor.server.testing.TestApplicationEngine
import io.ktor.server.testing.createTestEnvironment
import io.ktor.server.testing.handleRequest
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import junit.framework.Assert
import no.nav.sbl.dialogarena.mininnboks.ObjectMapperProvider
import no.nav.sbl.dialogarena.mininnboks.TestUtils
import no.nav.sbl.dialogarena.mininnboks.consumer.HenvendelseService
import no.nav.sbl.dialogarena.mininnboks.consumer.domain.*
import no.nav.sbl.dialogarena.mininnboks.consumer.tilgang.TilgangDTO
import no.nav.sbl.dialogarena.mininnboks.consumer.tilgang.TilgangService
import no.nav.sbl.dialogarena.mininnboks.provider.rest.ubehandletmelding.sporsmalController
import no.nav.tjeneste.domene.brukerdialog.henvendelse.v1.sendinnhenvendelse.meldinger.WSSendInnHenvendelseResponse
import org.apache.commons.lang3.StringUtils
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers.stringContainsInOrder
import org.hamcrest.core.Is
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import java.util.*
import java.util.stream.Collectors
import javax.ws.rs.BadRequestException
import javax.ws.rs.core.Response

class HenvendelseControllerTest : Spek({

    describe("Henvendelse tester") {

        val service = mockk<HenvendelseService>()
        val tilgangService  = mockk<TilgangService>()

        beforeEachTest {

            setUp(service, tilgangService)
        }

        val engine = TestApplicationEngine(createTestEnvironment())
        engine.start(wait = false) // for now we can't eliminate it

        engine.application.routing { HenvendelseController(service, tilgangService, false) }
        engine.application.install(ContentNegotiation) {
            register(ContentType.Application.Json, JacksonConverter(ObjectMapperProvider.objectMapper))
        }
        with(engine) {
          //  every { (henvendelseService.hentAlleHenvendelser(any())) } returns emptyList<Henvendelse>()

            it( "filtrerer Bort Uavsluttede Delsvar") {
                val lstHenvendelserBehandlingskjedeMedDelsvar = listOf(
                        Henvendelse("123").withType(Henvendelsetype.SPORSMAL_SKRIFTLIG).withTraadId("1").withOpprettetTid(TestUtils.now()),
                        Henvendelse("234").withType(Henvendelsetype.DELVIS_SVAR_SKRIFTLIG).withTraadId("1").withOpprettetTid(TestUtils.now())
                )
                every {service.hentAlleHenvendelser(any()) } returns lstHenvendelserBehandlingskjedeMedDelsvar
                handleRequest(HttpMethod.Get, "/traader") {
                }.apply {
                    MatcherAssert.assertThat(response.content, stringContainsInOrder(Henvendelsetype.DELVIS_SVAR_SKRIFTLIG.name))
                }

            }
        }
    }
})




   // @Rule
    //var subjectRule = SubjectRule(Subject("fnr", IdentType.EksternBruker, SsoToken.oidcToken("token", emptyMap<String, Any>())))

    fun setUp(service: HenvendelseService , tilgangService :TilgangService) {
        val henvendelser = Arrays.asList(
                Henvendelse("1").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(TestUtils.now()),
                Henvendelse("2").withTraadId("2").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(TestUtils.now()),
                Henvendelse("3").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(TestUtils.now()),
                Henvendelse("4").withTraadId("3").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(TestUtils.now()),
                Henvendelse("5").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(TestUtils.now()),
                Henvendelse("6").withTraadId("2").withType(Henvendelsetype.SPORSMAL_MODIA_UTGAAENDE).withOpprettetTid(TestUtils.nowPlus(100)),
                Henvendelse("7").withTraadId("1").withType(Henvendelsetype.SAMTALEREFERAT_OPPMOTE).withOpprettetTid(TestUtils.now())
        )

        val slot = slot<String>()
        every { service.hentAlleHenvendelser(any()) } returns henvendelser
        every { service.hentTraad(capture(slot)) } answers {
            val traadId = slot.captured
            henvendelser.stream()
                    .filter { henvendelse: Henvendelse? -> traadId == henvendelse!!.traadId }
                    .collect(Collectors.toList())
        }
        every {service.sendSvar(any(), any())} returns (
                WSSendInnHenvendelseResponse().withBehandlingsId(UUID.randomUUID().toString())
        )
    }

/*
    @Test
    @Throws(Exception::class)


    @Test
    @Throws(Exception::class)
   it("henter Ut Enkelt Traad Basert PaId") {
        val traad1 = controller.hentEnkeltTraad("1").entity as Traad
        val traad2 = controller.hentEnkeltTraad("2").entity as Traad
        val traad3 = controller.hentEnkeltTraad("3").entity as Traad
        MatcherAssert.assertThat(traad1.meldinger.size, Is.`is`(4))
        MatcherAssert.assertThat(traad2.meldinger.size, Is.`is`(2))
        MatcherAssert.assertThat(traad3.meldinger.size, Is.`is`(1))
    }

    @Test
    @Throws(Exception::class)
   it("henter Ut Traad Som Ikke Finnes") {
        val response = controller.hentEnkeltTraad("avabv")
        MatcherAssert.assertThat(response.status, Is.`is`(404))
    }

    @Test
   it("gir Statuskode Ikke Funnet Hvis Henvendelse Service Gir Soap Fault") {
        every {service!!.hentTraad(any()) } throws SOAPFaultException()
        val response = controller.hentEnkeltTraad("1")
        MatcherAssert.assertThat(response.status, Is.`is`(Response.Status.NOT_FOUND.statusCode))
    }

    @Test
    @Throws(Exception::class)
   it("markering Som Lest") {
        controller.markerSomLest("1")
        Mockito.verify(service, Mockito.times(1))?.merkSomLest("1")
    }

    @Test
    @Throws(Exception::class)
   it("markering Alle Som Lest") {
        controller.markerAlleSomLest("1")
        Mockito.verify(service, Mockito.times(1))?.merkAlleSomLest("1")
    }

    @Test
    @Throws(Exception::class)
   it("kan Ikke Sende Svar Nar Siste Henvendelse Ikke Er Sporsmal") {
        val svar = Svar()
        svar.traadId = "1"
        svar.fritekst = "Tekst"
        val response = controller.sendSvar(svar)
        MatcherAssert.assertThat(response.status, Is."is"(Response.Status.NOT_ACCEPTABLE.statusCode))
    }

    @Test
    @Throws(Exception::class)
   it("kan Sende Svar Nar Siste Henvendelse Er Sporsmal") {
        val svar = Svar()
        svar.traadId = "2"
        svar.fritekst = "Tekst"
        val nyHenvendelseResultat = controller.sendSvar(svar).entity as NyHenvendelseResultat
        MatcherAssert.assertThat(nyHenvendelseResultat.behandlingsId, Is.`is`(CoreMatchers.not(CoreMatchers.nullValue())))
    }

    @Test
   it("kopierer Nyeste Er Tilknyttet Ansatt Flagg Til Svaret") {
        val henvendelse1 = Henvendelse("1")
        henvendelse1.erTilknyttetAnsatt = true
        henvendelse1.opprettet = TestUtils.now()
        henvendelse1.type = Henvendelsetype.SPORSMAL_MODIA_UTGAAENDE
        val henvendelse2 = Henvendelse("2")
        henvendelse2.erTilknyttetAnsatt = false
        henvendelse2.opprettet = TestUtils.now()
        val henvendelser = Arrays.asList(henvendelse1, henvendelse2)
        every {service!!.hentTraad(ArgumentMatchers.anyString())} returns (henvendelser)
        val svar = Svar()
        svar.fritekst = "fritekst"
        svar.traadId = "0"
        controller.sendSvar(svar)
        val henvendelseArgumentCaptor = ArgumentCaptor.forClass(Henvendelse::class.java)
        Mockito.verify(service)?.sendSvar(henvendelseArgumentCaptor.capture(), ArgumentMatchers.anyString())
        MatcherAssert.assertThat(henvendelseArgumentCaptor.value.erTilknyttetAnsatt, Is."is"(true))
    }

    @Test
   it("kopierer Brukers Enhet Til Svaret") {
        val brukersEnhet = "1234"
        val henvendelse1 = Henvendelse("1")
        henvendelse1.opprettet = TestUtils.nowPlus(-1)
        henvendelse1.type = Henvendelsetype.SPORSMAL_SKRIFTLIG
        henvendelse1.brukersEnhet = brukersEnhet
        val henvendelse2 = Henvendelse("2")
        henvendelse2.opprettet = TestUtils.now()
        henvendelse2.type = Henvendelsetype.SPORSMAL_MODIA_UTGAAENDE
        every {service!!.hentTraad(anyString())} returns (Arrays.asList(henvendelse1, henvendelse2))
        val svar = Svar()
        svar.fritekst = "fritekst"
        svar.traadId = "0"
        controller.sendSvar(svar)
        val henvendelseArgumentCaptor = ArgumentCaptor.forClass(Henvendelse::class.java)
        Mockito.verify(service)?.sendSvar(henvendelseArgumentCaptor.capture(), ArgumentMatchers.anyString())
        MatcherAssert.assertThat(henvendelseArgumentCaptor.value.brukersEnhet, Is.`is`(brukersEnhet))
    }

    @Test
   it("sender Ikke Funnet Naar Traad Optional Ikke Er Present") {
        every {service!!.hentTraad(anyString())} returns (null)
        val svar = Svar()
        svar.fritekst = "fritekst"
        svar.traadId = "0"
        val response = controller.sendSvar(svar)
        MatcherAssert.assertThat(response.status, Is.`is`(Response.Status.NOT_FOUND.statusCode))
    }

    @Test(expected = BadRequestException::class)
   it("smeller Hvis Tom FritekstI Sporsmal") {
        val sporsmal = Sporsmal()
        sporsmal.fritekst = ""
        sporsmal.temagruppe = Temagruppe.ARBD.name
        controller.sendSporsmal(sporsmal)
    }

    @Test(expected = BadRequestException::class)
   it(`smeller Hvis For Lang Fritekst I Sporsmal") {
        val sporsmal = Sporsmal()
        sporsmal.fritekst = StringUtils.join(Collections.nCopies(1001, 'a'), "")
        sporsmal.temagruppe = Temagruppe.ARBD.name
        controller.sendSporsmal(sporsmal)
    }

    @Test(expected = BadRequestException::class)
   it("smeller Hvis Andre Sosial tjenester Temagruppe I Sporsmal") {
        val sporsmal = Sporsmal()
        sporsmal.fritekst = "DUMMY"
        sporsmal.temagruppe = Temagruppe.ANSOS.name
        controller.sendSporsmal(sporsmal)
    }

    @Test(expected = BadRequestException::class)
   it("smeller Hvis Bruker Er Kode6 Og Temagruppe OKSOS") {
        every {tilgangService.harTilgangTilKommunalInnsending(ArgumentMatchers.anyString())).thenReturn(
                TilgangDTO(TilgangDTO.Resultat.KODE6, "melding")
        )
        val sporsmal = Sporsmal()
        sporsmal.fritekst = "DUMMY"
        sporsmal.temagruppe = Temagruppe.OKSOS.name
        controller.sendSporsmal(sporsmal)
    }

    @Test(expected = BadRequestException::class)
   it("smeller Hvis Bruker Ikke Har Enhet Og Temagruppe OKSOS") {
        every {tilgangService.harTilgangTilKommunalInnsending(ArgumentMatchers.anyString())).thenReturn(
                TilgangDTO(TilgangDTO.Resultat.INGEN_ENHET, "melding")
        )
        val sporsmal = Sporsmal()
        sporsmal.fritekst = "DUMMY"
        sporsmal.temagruppe = Temagruppe.OKSOS.name
        controller.sendSporsmal(sporsmal)
    }

    @Test(expected = BadRequestException::class)
   it("smeller Hvis Temagruppe OKSOS Og Utledning Feiler") {
        every {tilgangService.harTilgangTilKommunalInnsending(ArgumentMatchers.anyString())).thenReturn(
                TilgangDTO(TilgangDTO.Resultat.FEILET, "melding")
        )
        val sporsmal = Sporsmal()
        sporsmal.fritekst = "DUMMY"
        sporsmal.temagruppe = Temagruppe.OKSOS.name
        controller.sendSporsmal(sporsmal)
    }

})*/

